### PR TITLE
Expose encountered header version in incompatibleHeader error

### PR DESCRIPTION
**Pull request by dfed**
This PR enables propagating the header version of the file being read as an associated value of the `CacheAdvanceError.incompatibleHeader` error. In order to do this, I've changed the signature of `func canOpenFile(...) -> Bool` to `func checkFile(...) throws` to push error throwing down the stack to where comparisons are being done and header version data is available.

Since we're adding new information to the incompatible header error, I'm considering this to be a feature version bump. This change is technically breaking to code that creates their own `incompatibleHeader` error, but since doing that outside of `CacheAdvance` doesn't make much sense I don't think this PR warrants a major version bump.

### CODE REVIEW

File: Sources/CacheAdvance/CacheHeaderHandle.swift
Comment (bachand) [inline Sources/CacheAdvance/CacheHeaderHandle.swift:80]: The name of this method makes me expect it will give me a file handle back. What about `checkFile()` in the spirit of this new similar Apple API ([source](https://github.com/DougGregor/swift-evolution/blob/fa58e0c4036ea3efbf4b264592aa99fb0f2f88f8/proposals/nnnn-structured-concurrency.md))?
Screen Shot 2021-01-11 at 5 26 43 PM

File: Sources/CacheAdvance/CacheAdvanceError.swift
Comment (bachand) [inline Sources/CacheAdvance/CacheAdvanceError.swift:24]: Let's document this associated value for completeness.

File: Sources/CacheAdvance/CacheHeaderHandle.swift
Comment (bachand) [inline Sources/CacheAdvance/CacheHeaderHandle.swift:150]:
```suggestion
    /// - Throws: `CacheAdvanceError.incompatibleHeader` if this object's header version does not match that of `fileHeader`.
```

Right?

File: Sources/CacheAdvance/CacheHeaderHandle.swift
Comment (bachand) [inline Sources/CacheAdvance/CacheHeaderHandle.swift:77]:
```suggestion
    /// Checks that the file is in a valid state ....<insert more?>....
```
Assuming that we go with the a name like `checkFile()` I wonder if we can make this more affirmative now and then keep the failure details in the `- Throws:` section?

File: Tests/CacheAdvanceTests/CacheHeaderHandleTests.swift
Comment (bachand) [inline Tests/CacheAdvanceTests/CacheHeaderHandleTests.swift:129]: Nice!

**Review by bachand — COMMENTED** [review]
@dfed I left some comments but as I finish up my review I'm having trouble seeing the key change in this PR. Can you help me follow? My read of the code indicates that the only change in the public API is that we propagate the currently persisted file header metadata version in the `CacheAdvanceError.incompatibleHeader` error (a good thing!). But from what I can tell we always propagated file read errors. What am I missing?

File: Sources/CacheAdvance/CacheAdvance.swift
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:76]: @bachand the changes in this file are the ones that cause the header-reading errors to be propagated. Previously we'd swallow whatever errors were encountered when calling `header.canOpenFile`. Now we propagate those errors directly.

Comment (bachand) [reply Sources/CacheAdvance/CacheAdvance.swift]: @dfed I feel silly posting this I feel like I must be missing something. I guess I'm not seeing what error we are swallowing. If `header.canOpenFile()` throws we propagate that error upward since the `guard` only applies if `header.canOpenFile()` returns normally.

Screen Shot 2021-01-11 at 8 42 09 PM

So I am pretty sure that if `memoizedMetadata()` (called within `canOpenFile()`) were to throw we would propagate that error outwards to the client.

What am I missing? 😖

Comment (dfed) [reply Sources/CacheAdvance/CacheAdvance.swift]: Good writeup @bachand! So... I think I was blinded by the fact because I'm seeing reports of `CacheAdvanceError.incompatibleHeader` being thrown and so I assumed we must be squashing a better error. If we're properly propagating errors encountered... then I have absolutely no clue how we're hitting this condition.

I'm open to reverting this part of the change, since you're right it doesn't make a functional difference. I do like the idea of `CacheHeaderHandle` throwing the error since it's the code doing the comparison, but you're right that doing so isn't a requirement.

Now I'm really curious what the version we're encountering is. If I had to guess: the versions will be somewhat random, and we're encountering files that shouldn't be where we're looking.

File: CacheAdvance.podspec
Comment (bachand) [inline CacheAdvance.podspec:3]: We're technically changing public API in the error enum case. But maybe that's small enough that it's OK?

File: Sources/CacheAdvance/CacheHeaderHandle.swift
Comment (bachand) [inline Sources/CacheAdvance/CacheHeaderHandle.swift:150]: @dfed I committed the change I proposed here since it seemed like you agreed.

File: Sources/CacheAdvance/CacheHeaderHandle.swift
Comment (bachand) [inline Sources/CacheAdvance/CacheHeaderHandle.swift:152]: Is this still true? If we only have a single version then how could we be hitting this problem?

**Review by bachand — APPROVED** [review]
@dfed I'm glad I understand now. I'm happy to move ahead since I think this is cleaner. Let's update the PR description before merging to clarify what we're actually changing.

File: Sources/CacheAdvance/CacheAdvance.swift
Comment (bachand) [reply Sources/CacheAdvance/CacheAdvance.swift]: Can you help me understand how the versions could be random? I'm not following that part.

File: Sources/CacheAdvance/CacheHeaderHandle.swift
Comment (dfed) [reply Sources/CacheAdvance/CacheHeaderHandle.swift:150]: Yes! Thank you for that 🙂

File: CacheAdvance.podspec
Comment (dfed) [reply CacheAdvance.podspec:3]: It's not a breaking change unless you're creating your own error... which... I don't think that's worth supporting with semver? A `catch CacheAdvanceError.incompatibleHeader` clause still works. Willing to roll a major version though since you're right that this is indeed a breaking change in some cases.

File: Sources/CacheAdvance/CacheHeaderHandle.swift
Comment (dfed) [reply Sources/CacheAdvance/CacheHeaderHandle.swift:152]: Good q! Will answer in this thread: https://github.com/dfed/CacheAdvance/pull/50#discussion_r555927694

File: Sources/CacheAdvance/CacheAdvance.swift
Comment (dfed) [reply Sources/CacheAdvance/CacheAdvance.swift]: The following code will throw a `CacheAdvanceError.incompatibleHeader(persistedVersion: 116)`:
```
FileManager.default.createFile(
    atPath: testFileLocation.path,
    contents: "this is a somewhat long string that is not real CacheAdvance data.".data(using: .utf8)!,
    attributes: nil)
try CacheHeaderHandle(
    forReadingFrom: testFileLocation,
    maximumBytes: 5000,
    overwritesOldMessages: false).checkFile()
```

To determine the header's version, we read the first `UInt8.storageLength` of whatever file a `CacheAdvance` instance is pointed at and turn that value into an integer. If we're reading a file that wasn't written by `CacheAdvance`, then the value of the version in the file "header" is effectively random/meaningless.

**Comment by dfed** [comment]
PR description updated! Waiting for explicit alignment on the versioning decision (feature bump vs breaking change) before merging.

File: CacheAdvance.podspec
Comment (bachand) [reply CacheAdvance.podspec:3]: I am OK keeping this a minor version since we wouldn't expect someone to be instantiating one of these errors.

File: Sources/CacheAdvance/CacheAdvance.swift
Comment (bachand) [reply Sources/CacheAdvance/CacheAdvance.swift]: Ah got it. We may want to consider adding a unique (or as-unique-as-possible) [signature](https://en.wikipedia.org/wiki/List_of_file_signatures) value to the start of the file. That could mitigate this problem significantly. If the file doesn't have the signature value we can bail out immediately and not try to parse it.

**Comment by bachand** [comment]
🚢

File: Sources/CacheAdvance/CacheAdvance.swift
Comment (dfed) [reply Sources/CacheAdvance/CacheAdvance.swift]: Yup! Thankfully we reserved 38bytes in our header, so we have room to encode a signature without making a breaking change. Though if we take this approach we'd need to enable an upgrade path for files written using the version without this info. Food for future thought.

In the meantime, the version matching is effectively a signature since we know all the possible valid values. It's a short and not-very-complex signature, but it's what we've got thus far 😅

Comment (dfed) [reply Sources/CacheAdvance/CacheAdvance.swift]: Interesting finding: the file size of the file with an incompatible header is _always_ 64bytes. Meaning that it's a file to which we never wrote any messages.

I still don't understand why the header file has an incompatible header (I screwed up and forgot to log the header version). But the good news is we're never losing messages.

**Pull request merged**
